controllers/workflow_controller.go

```diff
@@ -0,0 +1,454 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	wfcontroller = "workflow-controller"
+)
+
+type wfreconcile struct {
+	client       client.Client
+	log          logr.Logger
+	versionCache pkgaddon.VersionCacheClient
+}
+
+func NewWFController(mgr manager.Manager, stopChan <-chan struct{}, addonversioncache pkgaddon.VersionCacheClient) (controller.Controller, error) {
+	r := &wfreconcile{
+		client:       mgr.GetClient(),
+		log:          ctrl.Log.WithName(wfcontroller),
+		versionCache: addonversioncache,
+	}
+
+	c, err := controller.New(wfcontroller, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.Watch(&source.Kind{Type: &wfv1.Workflow{}}, r.enqueueRequestForOwner())
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *wfreconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.log.Info("reconciling", "request", req)
+	wfobj := &wfv1.Workflow{}
+	err := r.client.Get(ctx, req.NamespacedName, wfobj)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get workflow %s: %#v", req, err)
+	}
+
+	if len(string(wfobj.Status.Phase)) == 0 {
+		r.log.Info("workflow ", wfobj.GetNamespace(), wfobj.GetName(), " status", " is empty")
+		return ctrl.Result{}, nil
+	}
+
+	addonName, lifecycle, err := ExtractAddOnNameAndLifecycleStep(wfobj.GetName())
+	if err != nil {
+		msg := fmt.Sprintf("could not extract addon/lifecycle from %s/%s workflow.",
+			wfobj.GetNamespace(),
+			wfobj.GetName())
+		r.log.Info(msg)
+		return ctrl.Result{}, fmt.Errorf(msg)
+	}
+
+	err = r.updateAddonStatusLifecycle(ctx, wfobj.GetNamespace(), addonName, lifecycle, wfobj.Status.Phase)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *wfreconcile) enqueueRequestForOwner() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		var namespace = a.GetNamespace()
+		if namespace == workflowDeployedNS {
+			// Let's lookup addon related to this object.
+			return []reconcile.Request{
+				{NamespacedName: types.NamespacedName{
+					Name:      a.GetName(),
+					Namespace: namespace,
+				},
+				},
+			}
+		}
+		return []reconcile.Request{}
+	})
+}
+
+// extract addon-name and lifecyclestep from a workflow name string generated based on
+// api types
+func ExtractAddOnNameAndLifecycleStep(addonworkflowname string) (string, string, error) {
+	if strings.Contains(addonworkflowname, "prereqs") {
+		return strings.TrimSpace(addonworkflowname[:strings.Index(addonworkflowname, "prereqs")-1]), "prereqs", nil
```

Comment:
These strings like `prereqs` are already constants here:
https://github.com/keikoproj/addon-manager/blob/612ebb4d50bc8abb19392b5edcde18b8d7150e89/api/addon/v1alpha1/addon_types.go#L171-L180

Also this looks a little flaky, I think we said the Workflow has an owner ref for the Addon and if we really wanted to make this specific I think we should add an annotation the WF to push the specific lifecycle step, no extra parsing needed.

Reply:
Yes, we can leverage some existing definition, which is a valid points.
This is only pure string parse.